File: src/main/java/org/luxons/sevenwonders/game/boards/Board.java
package org.luxons.sevenwonders.game.boards;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;

import org.luxons.sevenwonders.game.Player;
import org.luxons.sevenwonders.game.Settings;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;
import org.luxons.sevenwonders.game.cards.Color;
import org.luxons.sevenwonders.game.effects.SpecialAbility;
import org.luxons.sevenwonders.game.resources.Production;
import org.luxons.sevenwonders.game.resources.TradingRules;
import org.luxons.sevenwonders.game.scoring.PlayerScore;
import org.luxons.sevenwonders.game.scoring.ScoreCategory;
import org.luxons.sevenwonders.game.wonders.Wonder;

public class Board {

    private final Wonder wonder;

    private final Player player;

    private final List<Card> playedCards = new ArrayList<>();

    private final Production production = new Production();

    private final Science science = new Science();

    private final TradingRules tradingRules;

    private final Military military;

    private final Set<SpecialAbility> specialAbilities = EnumSet.noneOf(SpecialAbility.class);



    private int gold;

    public Board(Wonder wonder, Player player, Settings settings) {
        this.wonder = wonder;
        this.player = player;
        this.gold = settings.getInitialGold();
        this.tradingRules = new TradingRules(settings.getDefaultTradingCost());
        this.military = new Military(settings);
        production.addFixedResource(wonder.getInitialResource(), 1);
    }

    public Wonder getWonder() {
        return wonder;
    }

    public Player getPlayer() {
        return player;
    }

    public List<Card> getPlayedCards() {
        return playedCards;
    }

    public void addCard(Card card) {
        playedCards.add(card);
    }

    public int getNbCardsOfColor(List<Color> colorFilter) {
        return (int) playedCards.stream().filter(c -> colorFilter.contains(c.getColor())).count();
    }

    public boolean isPlayed(String cardName) {
        return getPlayedCards().stream().map(Card::getName).filter(name -> name.equals(cardName)).count() > 0;
    }

    public Production getProduction() {
        return production;
    }

    public TradingRules getTradingRules() {
        return tradingRules;
    }

    public Science getScience() {
        return science;
    }

    public int getGold() {
        return gold;
    }

    public void setGold(int amount) {
        this.gold = amount;
    }

    public void addGold(int amount) {
        this.gold += amount;
    }

    public void removeGold(int amount) {
        if (gold < amount) {
            throw new InsufficientFundsException(gold, amount);
        }
        this.gold -= amount;
    }

    public Military getMilitary() {
        return military;
    }

    public void addSpecial(SpecialAbility specialAbility) {
        specialAbilities.add(specialAbility);
    }

    public boolean hasSpecial(SpecialAbility specialAbility) {
        return specialAbilities.contains(specialAbility);
    }

    public PlayerScore computePoints(Table table) {
        PlayerScore score = new PlayerScore(player, gold);
        score.put(ScoreCategory.CIVIL, computePointsForCards(table, Color.BLUE));
        score.put(ScoreCategory.MILITARY, military.getTotalPoints());
        score.put(ScoreCategory.SCIENCE, science.computePoints());
        score.put(ScoreCategory.TRADE, computePointsForCards(table, Color.YELLOW));
        score.put(ScoreCategory.GUILD, computePointsForCards(table, Color.PURPLE));
        score.put(ScoreCategory.WONDER, wonder.computePoints(table, player.getIndex()));
        return score;
    }

    private int computePointsForCards(Table table, Color color) {
        return playedCards.stream()
                .filter(c -> c.getColor() == color)
                .flatMap(c -> c.getEffects().stream())
                .mapToInt(e -> e.computePoints(table, player.getIndex()))
                .sum();
    }

    static class InsufficientFundsException extends RuntimeException {
        InsufficientFundsException(int current, int required) {
            super(String.format("Current balance is %d gold, but %d are required", current, required));
        }
    }
}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/engine/boards/Board.kt
package org.luxons.sevenwonders.engine.boards

import org.luxons.sevenwonders.engine.Player
import org.luxons.sevenwonders.engine.Settings
import org.luxons.sevenwonders.engine.cards.Card
import org.luxons.sevenwonders.engine.effects.SpecialAbility
import org.luxons.sevenwonders.engine.resources.Production
import org.luxons.sevenwonders.engine.resources.TradingRules
import org.luxons.sevenwonders.engine.resources.mutableResourcesOf
import org.luxons.sevenwonders.engine.wonders.Wonder
import org.luxons.sevenwonders.model.Age
import org.luxons.sevenwonders.model.cards.Color
import org.luxons.sevenwonders.model.score.PlayerScore
import org.luxons.sevenwonders.model.score.ScoreCategory

internal class Board(val wonder: Wonder, val playerIndex: Int, settings: Settings) {

    val production = Production(mutableResourcesOf(wonder.initialResource))
    val publicProduction = Production(mutableResourcesOf(wonder.initialResource))
    val science = Science()
    val military: Military = Military(settings.lostPointsPerDefeat, settings.wonPointsPerVictoryPerAge)
    val tradingRules: TradingRules = TradingRules(settings.defaultTradingCost)

    private val pointsPer3Gold: Int = settings.pointsPer3Gold

    private val playedCards: MutableList<Card> = mutableListOf()
    private val specialAbilities: MutableSet<SpecialAbility> = hashSetOf()
    private val consumedFreeCards: MutableMap<Age, Boolean> = mutableMapOf()

    var gold: Int = settings.initialGold
        private set

    var copiedGuild: Card? = null
        internal set(copiedGuild) {
            if (copiedGuild!!.color !== Color.PURPLE) {
                throw IllegalArgumentException("The given card '$copiedGuild' is not a Guild card")
            }
            field = copiedGuild
        }

    fun getPlayedCards(): List<Card> = playedCards

    fun addCard(card: Card) {
        playedCards.add(card)
    }

    fun getNbCardsOfColor(colorFilter: List<Color>): Int = playedCards.count { colorFilter.contains(it.color) }

    fun isPlayed(cardName: String): Boolean = playedCards.count { it.name == cardName } > 0

    fun addGold(amount: Int) {
        this.gold += amount
    }

    fun removeGold(amount: Int) {
        if (gold < amount) {
            throw InsufficientFundsException(gold, amount)
        }
        this.gold -= amount
    }

    fun addSpecial(specialAbility: SpecialAbility) {
        specialAbilities.add(specialAbility)
    }

    fun removeSpecial(specialAbility: SpecialAbility) {
        specialAbilities.remove(specialAbility)
    }

    fun hasSpecial(specialAbility: SpecialAbility): Boolean = specialAbilities.contains(specialAbility)

    fun canPlayFreeCard(age: Age): Boolean =
        hasSpecial(SpecialAbility.ONE_FREE_PER_AGE) && !consumedFreeCards.getOrDefault(age, false)

    fun consumeFreeCard(age: Age) {
        consumedFreeCards[age] = true
    }

    fun computeScore(player: Player): PlayerScore = PlayerScore(
        playerIndex = playerIndex,
        boardGold = gold,
        pointsByCategory = mapOf(
            ScoreCategory.CIVIL to computePointsForCards(player, Color.BLUE),
            ScoreCategory.MILITARY to military.totalPoints,
            ScoreCategory.SCIENCE to science.computePoints(),
            ScoreCategory.TRADE to computePointsForCards(player, Color.YELLOW),
            ScoreCategory.GUILD to computePointsForCards(player, Color.PURPLE),
            ScoreCategory.WONDER to wonder.computePoints(player),
            ScoreCategory.GOLD to computeGoldPoints()
        )
    )

    private fun computePointsForCards(player: Player, color: Color): Int =
        playedCards.filter { it.color === color }
            .flatMap { it.effects }
            .sumBy { it.computePoints(player) }

    private fun computeGoldPoints(): Int = gold / 3 * pointsPer3Gold

    internal class InsufficientFundsException(current: Int, required: Int) :
        IllegalStateException("Current balance is $current gold, but $required are required")
}
